imagetagger/imagetagger/images/models.py

```diff
@@ -68,10 +84,61 @@ class Meta:
         default=None
     )
     pinned_by = models.ManyToManyField(settings.AUTH_USER_MODEL, related_name='pinned_sets')
+    zip_finished = models.BooleanField(default=True)
 
     def root_path(self):
         return os.path.join(settings.IMAGE_PATH, self.path)
 
+    def zip_file_exists(self):
+        return os.path.isfile(self.zip_path())
+
+    def zip_path(self):
+        return os.path.join(settings.IMAGE_PATH, self.relative_zip_path())
+
+    def relative_zip_path(self):
+        return os.path.join(self.path, self.name) + '.zip'
+
+    def _update_zip_thread(self):
+        self.zip_finished = False
+        self.save()
+
+        # The python standard library module 'zipfile' is used for
+        # the zipping process. It allows adding of files with a
+        # different file name in the zip file which is crucial because
+        # the internal store process adds random strings to the
+        # filename that should not appear in the zip archive.
+        # Unfortunately, the module does not support deletion of
+        # single files in the archive. Therefore, the 'zip' command
+        # line tool is called as a subprocess.
+
+        with zipfile.ZipFile(self.zip_path(), 'a') as f:
+            contents = f.namelist()
+            # Add missing image to the archive
+            for image in self.images.all():
+                if image.name in contents:
+                    contents.remove(image.name)
+                else:
+                    f.write(image.path(), image.name)
+
+        # Remove deleted images
+        if len(contents) > 0:
+            command = ['zip', '-d', self.zip_path()] + contents
+            subprocess.run(command)
+
+        self.zip_finished = True
+        self.save()
+
+    def update_zip(self, threading=True):
+        # Threading is used to avoid interrupting the user for the
+        # potentially long zip file creation. Sometimes though, the
+        # user should have to wait (e. g. for a download), hence the
+        # 'threading' parameter.
+        if threading:
+            t = Thread(target=self._update_zip_thread)
+            t.start()
```

Comment:
I am not sure if it is a good idea to start a Thread in a web worker, as it could block the worker process. This has to be tested with uwsgi. What happens if the worker gets scraped?

Reply:
I too feel rather uncomfortable with using a thread here. The zip files should probably better be created and updated in a background process (management command); it is not the use case of the web workers to do such long-running tasks (and especially not in a thread that is just disregarded at the end of the request!)

Reply:
http://django-background-tasks.readthedocs.io/en/latest/ seems to be a good way to do background processing